group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/GeneralUniformAssignmentBuilder.java

```diff
@@ -14,17 +14,891 @@
+
+        if (rackInfo.useRackStrategy) rackAwarePartitionAssignment();
+        unassignedPartitionsAssignment();
+
+        balance();
+
+        return new GroupAssignment(targetAssignment);
+    }
+
+    /**
+     * <li> TopicIdPartitions are sorted in descending order based on the value:
+     *       totalPartitions/number of subscribed members. </li>
+     * <li> If the above value is the same then topicIdPartitions are sorted in
+     *      ascending order of number of subscribers. </li>
+     * <li> If both criteria are the same, sort in ascending order of the partition Id.
+     *      This last criteria is for predictability of the assignments. </li>
+     *
+     * @param topicIdPartitions       The topic partitions that need to be sorted.
+     * @return A list of sorted topic partitions.
+     */
+    private List<TopicIdPartition> sortTopicIdPartitions(Collection<TopicIdPartition> topicIdPartitions) {
+        Comparator<TopicIdPartition> comparator = Comparator
+            .comparingDouble((TopicIdPartition topicIdPartition) -> {
+                int totalPartitions = subscribedTopicDescriber.numPartitions(topicIdPartition.topicId());
+                int totalSubscribers = membersPerTopic.get(topicIdPartition.topicId()).size();
+                return (double) totalPartitions / totalSubscribers;
+            })
+            .reversed()
+            .thenComparingInt(topicIdPartition -> membersPerTopic.get(topicIdPartition.topicId()).size())
+            .thenComparingInt(TopicIdPartition::partitionId);
+
+        return topicIdPartitions.stream()
+            .sorted(comparator)
+            .collect(Collectors.toList());
+    }
+
+    /**
+     * Gets a set of partitions that are to be retained from the existing assignment. This includes:
+     * <li> Partitions from topics that are still present in both the new subscriptions and the topic metadata. </li>
+     * <li> When using a rack-aware strategy, only partitions with member owners in the same rack are retained. </li>
+     * <li> Track current partition owners when there is a rack mismatch. </li>
+     */
+    private void assignStickyPartitions() {
+        members.forEach((memberId, assignmentMemberSpec) ->
+            assignmentMemberSpec.assignedPartitions().forEach((topicId, currentAssignment) -> {
+                if (assignmentMemberSpec.subscribedTopicIds().contains(topicId)) {
+                    currentAssignment.forEach(partition -> {
+                        TopicIdPartition topicIdPartition = new TopicIdPartition(topicId, partition);
+                        if (rackInfo.useRackStrategy && rackInfo.racksMismatch(memberId, topicIdPartition)) {
+                            currentPartitionOwners.put(topicIdPartition, memberId);
+                        } else {
+                            assignmentManager.maybeAssignPartitionToMember(topicIdPartition, memberId);
+                            assignedStickyPartitions.add(topicIdPartition);
```

Comment:
yeah that's a valid concern, let me write a test for this and think about it, but we can probably add the check to be safe anyways. I think the conditions that we check for in maybeAssignPartitionsToMember are unlikely to be satisfied for this particular case that's why it wasn't checked.

Reply:
So the two conditions in maybeAssignPartitionsToMember are a) is the member subscribed to the topic in question b) is the member at max capacity. The first condition cannot be true since we check for topic subscriptions before assigning sticky partitions. The second condition cannot be true either since max assignment refers to the sum total of all the subscribed partitions, which means all the sticky partitions would be assigned to the member anyways. I'm going to change it to directly assign partition to member in this case